components/once/once.c

```diff
@@ -0,0 +1,30 @@
+#include "once.h"
+
+#include <stdatomic.h>
+#include <stdbool.h>
+#include <uv.h>
+
+int datadog_php_once_ctor(struct datadog_php_once_s *once) {
+  atomic_store(&once->initialized, false);
+  return uv_mutex_init(&once->mutex);
+}
+
+void datadog_php_once_dtor(struct datadog_php_once_s *once) {
+  atomic_store(&once->initialized, true);
+  uv_mutex_destroy(&once->mutex);
+}
+
+void datadog_php_once(struct datadog_php_once_s *once,
+                      void (*init_routine)(void)) {
+  // If it's already initialized we can avoid the lock.
+  if (atomic_load(&once->initialized)) {
+    return;
+  }
+  uv_mutex_lock(&once->mutex);
```

Comment:
if you destroy while the lock is acquired, are things not going badly ?

Reply:
Destructor and constructor are not thread-safe. I documented this in the header which will work fine for the intended use-case.